# Team Justus Von Liebug

## Introduction: *Minimum Cost Diet*s for Diabetic Undergraduate Students at UC Berkeley.

*“Prediabetes is more common than you may realize – one in three adults has prediabetes and 90% don’t know it.  At Berkeley, an estimated 28% of faculty and staff have prediabetes and this is a 29% increase over a two year time period.”* - UC Berkeley Be Well at Work Program

Diabetes affects all demographics but disproportionately affects low income communities and adults where the risk increases after adulthood. As of lately, managing diabetes has been a topic among Berkeley policy and University programs to help mitigate long-term health effects of diabetes. A subsistence cost diet would be an effective way in addition to policy and regulation to make managing diabetes sustiainable and effective for low income individuals, especially students!


We begin by installing material from lecture.

In [1]:
!pip install -r requirements.txt;
import numpy as np
import pandas as pd
import fooddatacentral as fdc

In [2]:
!gpg --batch --passphrase "importance variation" -d justus_von_liebig-069d38b622a8.json.gpg > justus_von_liebig-069d38b622a8.json

gpg: AES encrypted data
gpg: encrypted with 1 passphrase


User and API Key from USDA:

In [3]:
user = "ligon"

# API key for Gov; substitute your own!
apikey = "NVgmRkqQ7ikdgus7M2OTqZFlDXm67vdQ3aiP1GND"

# File with private keys for relevant service account to authenticate
# and access google spreadsheets
serviceacct = {'ligon':'students-9093fa174318.json'}

## Data Collection

Using previous purchases of our group members from receipts, we compiled a small database of groceries we assume to be available to a typical UC Berkeley undergraduate. This data was recorded and stored in a Google Sheet. The URL can be found here: https://docs.google.com/spreadsheets/d/1xnjnl42g_jYKdkOfW2OF7tBdGmuCAeJluT2O7fx1OQE/edit?usp=sharing

Information from the following categories were inputted:

- **Food:** Label used to identify food
- **Quantity:** Amount of food
- **Units:** Units in which amounts of this food is measured
- **Price:** Price for quantity of food
- **FDC Code** Key identifying food in the FoodDataCentral dataset. *See below.*

Food from our sample was manually matched to foods in the FDC database using the following code:

In [4]:
import fooddatacentral as fdc

# Search by typing the name of the food.
fdc.search(apikey,"green bell peppers")

fdcId                                        description  \
0   1103561                             Peppers, green, cooked   
1   1104580                Green pepper, cooked, as ingredient   
2   1103371                          Pepper, sweet, green, raw   
3    168577                      Peppers, chili, green, canned   
4    170427                         Peppers, sweet, green, raw   
5    170024                     Peppers, sweet, green, sauteed   
6    170497                     Peppers, hot chili, green, raw   
7    170336                                  TACO BELL, Nachos   
8    170429  Peppers, sweet, green, canned, solids and liquids   
9    170022  Peppers, sweet, green, frozen, chopped, unprep...   
10  1099472           Stuffed green pepper, Puerto Rican style   
11   169218  Corn with red and green peppers, canned, solid...   
12   170109  Peppers, sweet, green, cooked, boiled, drained...   
13   171605  Sauce, chili, peppers, hot, immature green, ca...   
14   170734                            TACO BELL, Bean Burrito   
15   170337                          TACO BELL, Nachos Supreme   
16   170736                              TACO BELL, Taco Salad   
17   170428  Peppers, sweet, green, cooked, boiled, drained...   
18   170462  Tomato products, canned, sauce, with onions, g...   
19   170735               TACO BELL, BURRITO SUPREME with beef   
20   170334            TACO BELL, BURRITO SUPREME with chicken   
21   170335              TACO BELL, BURRITO SUPREME with steak   
22   170733                    TACO BELL, Soft Taco with steak   
23   170023  Peppers, sweet, green, frozen, chopped, boiled...   
24   170516  Peppers, sweet, green, frozen, chopped, cooked...   
25   170426  Peppers, hot chili, green, canned, pods, exclu...   
26  1103684                                   Peppers, pickled   
27   170332  TACO BELL, Original Taco with beef, cheese and...   
28   170333  TACO BELL, Soft Taco with beef, cheese and let...   
29   170732  TACO BELL, Soft Taco with chicken, cheese and ...   
30  1099608                                       Pepper steak   
31  1103679                                      Olives, green   
32  1103685                               Pepper, hot, pickled   
33  1103373                                Pepper, banana, raw   
34  1103370                                   Pepper, raw, NFS   
35  1103562                               Peppers, red, cooked   
36  1103629                               Green bean casserole   
37  1103282                               Fried green tomatoes   
38  1103683                                   Hot pepper sauce   
39  1103563                                Hot peppers, cooked   
40   169394                                Pepper, banana, raw   
41  1103928                      Pepper, for use on a sandwich   
42  1103368                               Pepper, poblano, raw   
43  1103369                               Pepper, Serrano, raw   
44   169396                              Peppers, ancho, dried   
45   168578                            Peppers, hungarian, raw   
46   168576                             Peppers, jalapeno, raw   
47   168579                            Peppers, pasilla, dried   
48   169395                              Peppers, serrano, raw   
49  1099640                      Sausage and peppers, no sauce   

                                 lowercaseDescription commonNames  \
0                              peppers, green, cooked               
1                 green pepper, cooked, as ingredient               
2                           pepper, sweet, green, raw               
3                       peppers, chili, green, canned               
4                          peppers, sweet, green, raw               
5                      peppers, sweet, green, sauteed               
6                      peppers, hot chili, green, raw               
7                                   taco bell, nachos               
8   peppers, sweet

## Loading Groceries Data from Google Sheets

The following loads data inputted in Google Sheets and creates a data frame.

Creating dataframe containing 'Groceries' data:

In [5]:
df = pd.read_csv('groceries.csv')
df = df[df['Units'].isin(['g', 'lb'])]
df.sort_values('Price')

Food  Quantity Units  Price  \
41                              Fair trade bananas      1.00    lb   1.19   
39                               Org green cabbage      1.00    lb   1.29   
33                             Org bunched spinach      1.00    lb   1.29   
32                                     Org carrots      1.00    lb   1.29   
29                               Org yellow onions      1.00    lb   1.49   
40                                 Org Fuji apples      1.00    lb   1.69   
58                 Dole, Organic Bananas, 2 lb Bag      2.00    lb   1.69   
34                     Org Large yukon gold potato      1.00    lb   1.69   
48                boneless skinless chicken breast      1.00    lb   1.84   
30                                    Org zucchini      1.00    lb   2.29   
57                         Lemons, Small, 2 lb bag      2.00    lb   2.40   
47                                     ground beef      1.00    lb   2.44   
27                              Org Steel cut oats      1.80    lb   2.99   
66                        California sweet Oranges      3.00    lb   3.19   
59              Wonderful Halos Mandarins, 3lb Bag      3.00    lb   3.77   
68                                       Asparagus      2.00    lb   3.98   
4                                      Sanuki Udon      2.45    lb   3.99   
42                                    Org mandarin      2.00    lb   3.99   
8                                     Napa cabbage      4.99    lb   4.94   
11                              Rotisserie Chicken      3.00    lb   4.99   
67                                Fish cake Frozen      1.00    lb   4.99   
3                                  Sajo Light Tuna    150.00     g   5.00   
24                boneless skinless chicken breast      3.00    lb   5.94   
51  Natural Fresh Chicken Thighs | Antibiotic Free      1.25    lb   5.99   
2                           Ottogi Gold Mayonnaise    500.00     g   5.99   
36                       Challenge unsalted butter      1.00    lb   5.99   
35                             Org large mushrooms      1.00    lb   6.50   
64                                 chicken breast       2.50    lb   6.59   
9                              Crushed red pepper       1.00    lb   6.95   
65                               Fish Filet frozen      1.50    lb   6.99   
7                         pork belly skinless thin      0.92    lb   7.91   
12                                Broccoli Florets      1.00    lb   8.99   
44                                    Cream Cheese      3.00    lb   9.09   
6                         Pork belly skin on thick      1.10    lb   9.81   
38                                    Organic rice     10.00    lb  10.49   
46                                  cheddar cheese      5.00    lb  11.99   
10                                   Salted shrimp      1.10    lb  11.99   
50                              green bell peppers     25.00    lb  18.79   
49                              Lean ground turkey      6.68    lb  20.59   

        FDC  
41   407274  
39  1103341  
33  1377964  
32  1286361  
29   790646  
40  1414072  
58   407274  
34  1274336  
48  1332817  
30   595040  
57  1102594  
47  1253531  
27  1252490  
66  1102597  
59  1369644  
68  1383906  
4    548153  
42   169105  
8   1276159  
11  1279544  
67  1099380  
3    175160  
24  1266762  
51  1451906  
2   1103879  
36   731258  
35  1265847  
64  1328384  
9   1186269  
65   333374  
7    167812  
12  1255048  
44  1204868  
6    167812  
38   356554  
46  1250829  
10  1257419  
50  1103371  
49  1255363

As different items in the 'groceries' dataset are listed at different price points for different quantities and units, we normalize the data to appear similar for comparison and analysis. The FDC database has a preference for items in terms of hundreds of grams (hectograms) or hundreds of milliliters (deciliters).

The units function is used to convert all foods to either deciliters or hectograms, to match FDC database:

In [116]:
# Convert food quantities to FDC units
df['FDC Quantity'] = df[['Quantity','Units']].T.apply(lambda x : fdc.units(x['Quantity'],x['Units']))

# Now divide price by the FDC Quantity to get, e.g., price per hectoliter
df['FDC Price'] = df['Price']/df['FDC Quantity']
df.dropna(how='any') # Drop food with any missing data

df

Food  Quantity Units  Price  \
2                           Ottogi Gold Mayonnaise    500.00     g   5.99   
3                                  Sajo Light Tuna    150.00     g   5.00   
4                                      Sanuki Udon      2.45    lb   3.99   
6                         Pork belly skin on thick      1.10    lb   9.81   
7                         pork belly skinless thin      0.92    lb   7.91   
8                                     Napa cabbage      4.99    lb   4.94   
9                              Crushed red pepper       1.00    lb   6.95   
10                                   Salted shrimp      1.10    lb  11.99   
11                              Rotisserie Chicken      3.00    lb   4.99   
12                                Broccoli Florets      1.00    lb   8.99   
24                boneless skinless chicken breast      3.00    lb   5.94   
27                              Org Steel cut oats      1.80    lb   2.99   
29                               Org yellow onions      1.00    lb   1.49   
30                                    Org zucchini      1.00    lb   2.29   
32                                     Org carrots      1.00    lb   1.29   
33                             Org bunched spinach      1.00    lb   1.29   
34                     Org Large yukon gold potato      1.00    lb   1.69   
35                             Org large mushrooms      1.00    lb   6.50   
36                       Challenge unsalted butter      1.00    lb   5.99   
38                                    Organic rice     10.00    lb  10.49   
39                               Org green cabbage      1.00    lb   1.29   
40                                 Org Fuji apples      1.00    lb   1.69   
41                              Fair trade bananas      1.00    lb   1.19   
42                                    Org mandarin      2.00    lb   3.99   
44                                    Cream Cheese      3.00    lb   9.09   
46                                  cheddar cheese      5.00    lb  11.99   
47                                     ground beef      1.00    lb   2.44   
48                boneless skinless chicken breast      1.00    lb   1.84   
49                              Lean ground turkey      6.68    lb  20.59   
50                              green bell peppers     25.00    lb  18.79   
51  Natural Fresh Chicken Thighs | Antibiotic Free      1.25    lb   5.99   
57                         Lemons, Small, 2 lb bag      2.00    lb   2.40   
58                 Dole, Organic Bananas, 2 lb Bag      2.00    lb   1.69   
59              Wonderful Halos Mandarins, 3lb Bag      3.00    lb   3.77   
64                                 chicken breast       2.50    lb   6.59   
65                               Fish Filet frozen      1.50    lb   6.99   
66                        California sweet Oranges      3.00    lb   3.19   
67                                Fish cake Frozen      1.00    lb   4.99   
68                                       Asparagus      2.00    lb   3.98   

        FDC                  FDC Quantity                        FDC Price  
2   1103879                 5.0 hectogram                1.198 / hectogram  
3    175160                 1.5 hectogram   3.3333333333333335 / hectogram  
4    548153  11.113013065000002 hectogram   0.3590385412725149 / hectogram  
6    167812   4.989516070000001 hectogram   1.9661225382124077 / hectogram  
7    167812   4.173049804000001 hectogram   1.8954961890025885 / hectogram  
8   1276159  22.634259263000004 hectogram  0.21825322148162227 / hectogram  
9   1186269   4.535923700000001 hectogram   1.5322127221848991 / hectogram  
10  1257419   4.989516070000001 hectogram    2.403038657815165 / hectogram  
11  1279544          13.6077711 hectogram   0.3667022294341797 / hectogram  
12  1255048   4.535923700000001 hectogram   1.9819557370420493 / hectogram  
24  1266762          13.6077711 hectogram   0.4365152791260576 / hectogram  
27  1252490   8.164662660000001 hectogram

In [117]:
# To use minimum price observed
Prices = df.groupby('Food')['FDC Price'].min()
Prices

Food
Asparagus                                          0.4387199017479063 / hectogram
Broccoli Florets                                   1.9819557370420493 / hectogram
California sweet Oranges                          0.23442487212325314 / hectogram
Challenge unsalted butter                          1.3205689504874165 / hectogram
Cream Cheese                                        0.668000654420179 / hectogram
Crushed red pepper                                 1.5322127221848991 / hectogram
Dole, Organic Bananas, 2 lb Bag                   0.18629061154622154 / hectogram
Fair trade bananas                                0.26235009200000425 / hectogram
Fish Filet frozen                                  1.0273541417815295 / hectogram
Fish cake Frozen                                    1.100106688302539 / hectogram
Lean ground turkey                                 0.6795386195189564 / hectogram
Lemons, Small, 2 lb bag                           0.26455471462185304 / hectogram
Napa cabbag

## Mapping Food to Nutrients

Using our dataset of groceries, we look up nutritional values of foods that are accessible and consumed by UC Berkeley undergraduate students. 

In [7]:
# Creates a new dataframe detailing nutrients of items in 'groceries' dataset

import fooddatacentral as fdc

D = {}
for food in df['Food'].tolist():
    try:
        fdc_id = int((df[df['Food'] == food].FDC).iloc[0])
        D[food] = fdc.nutrients(apikey, fdc_id).Quantity
    except AttributeError: 
        warnings.warn("Couldn't find FDC Code %s for food %s." % (food, FDC))

FoodNutrients = pd.DataFrame(D,dtype=float)
FoodNutrients

Ottogi Gold Mayonnaise  Sajo Light Tuna  \
10:0                                               0.000              NaN   
11:0                                                 NaN              NaN   
12:0                                               0.000              NaN   
14:0                                               0.055            0.014   
14:1 c                                               NaN              NaN   
...                                                  ...              ...   
Vitamin K (Dihydrophylloquinone)                     NaN              NaN   
Vitamin K (phylloquinone)                        163.000              NaN   
Vitamins and Other Components                      0.000            0.000   
Water                                             21.650           63.160   
Zinc, Zn                                           0.150            0.560   

                                  Sanuki Udon  Pork belly skin on thick  \
10:0                                      NaN                      0.04   
11:0                                      NaN                       NaN   
12:0                                      NaN                      0.06   
14:0                                      NaN                      0.70   
14:1 c                                    NaN                       NaN   
...                                       ...                       ...   
Vitamin K (Dihydrophylloquinone)          NaN                      0.00   
Vitamin K (phylloquinone)                 NaN                      0.00   
Vitamins and Other Components             NaN                      0.00   
Water                                     NaN                     36.74   
Zinc, Zn                                  NaN                      1.02   

                                  pork belly skinless thin  Napa cabbage  \
10:0                                                  0.04           NaN   
11:0                                                   NaN           NaN   
12:0                                                  0.06           NaN   
14:0                                                  0.70           NaN   
14:1 c                                                 NaN           NaN   
...                                                    ...           ...   
Vitamin K (Dihydrophylloquinone)                      0.00           NaN   
Vitamin K (phylloquinone)                             0.00           NaN   
Vitamins and Other Components                         0.00           NaN   
Water                                                36.74           NaN   
Zinc, Zn                                              1.02           NaN   

                                  Crushed red pepper   Salted shrimp  \
10:0                                              NaN            NaN   
11:0                                              NaN            NaN   
12:0                                              NaN            NaN   
14:0                                              NaN            NaN   
14:1 c                                            NaN            NaN   
...                                               ...            ...   
Vitamin K (Dihydrophylloquinone)                  NaN            NaN   
Vitamin K (phylloquinone)                         NaN            NaN   
Vitamins and Other Components                     NaN            NaN   
Water                                             NaN            NaN   
Zinc, Zn                                          NaN            NaN   

                                  Rotisserie Chicken  Broccoli Florets  ...  \
10:0                                             NaN               NaN  ...   
11:0                                             NaN               NaN  ...   
12:0                                             NaN               NaN  ...   
14:0                                             NaN               NaN  ...   
14:1 c                             

## Dietary Requirements for Diabetes

Although there is no one-size-fits-all approach that works for every individual with diabetes, research reveals a general trend  in diets for those with diabetes to manage blood sugar levels and ultimately maintain long-term health. The American Diabetes Association (ADA) recommends a diet that does the following for diabetics:

1. Reduces carbohydrates intake
2. Manages fat intake (reduce high saturated fats and trans fats)
3. Limits sodium to less than 2,300 mg a day
4. Increases fiber intake


We reference US government recommendations available at https://www.dietaryguidelines.gov/sites/default/files/2019-05/2015-2020_Dietary_Guidelines.pdf according to age and sex. We then make additional adjustments to these recommendations for someone with diabetes.

**Sources:**
https://www.diabetes.org/healthy-living/recipes-nutrition
https://www.diabetesfoodhub.org/articles/what-is-the-diabetes-plate-method.html#:~:text=The%20Diabetes%20Plate%20Method%20is,you%20need%20is%20a%20plate!
https://www.mayoclinic.org/diseases-conditions/diabetes/in-depth/diabetes-diet/art-20044295

## 1. A Low-Carbohydrate/Low-Fat Diet

**Carbohydrates**

We define a low-carbohydrate diet to be less than recommended intake of 130 grams of carbohydrates a day (exlcuding fiber). In this case, diabetics does not need to drastically cut carbohydrate intake, as long as they have a wholistically balanced diet. Ideally, carbohydrate consumption would be approximately 30-50% of the recommended daily carbohydrate goals of the average person. This translates to about 40 g to 70 g/day. In this project, we limit carbohydrate intake to 50 grams a day.

**Fats**

The DRI for fat in adults is 20-35% of total calories from fat. American Diabetes Association (ADA) follows this recommendation for diabetic diets as well! For a 2000 calorie diet, this is about 44 to 77 g of fat per day. Furthermore, the ADA recommends avoiding saturated fats, which should make up 3-7% of total calories per day.

In [35]:
# Load minimum dietary requirements
bmin = pd.read_csv('./diet_minimums.csv').set_index('Nutrition').iloc[:, 2:]

# Add dietary minimum for carbohydrate
bmin.loc['Carbohydrate, by difference'] = 50
bmin

# Add dietary minimum for total fat
bmin.loc[len(bmin.index)] = np.around(bmin.loc['Energy'].values * 0.20 / 9)
bmin = bmin.rename({len(bmin.index)-1: 'Total fat (NLEA)'})

# Add dietary minimum for saturated fat
bmin.loc[len(bmin.index)] = np.around(bmin.loc['Energy'].values * 0.03 / 9)
bmin = bmin.rename({len(bmin.index)-1: 'Fatty acids, total saturated'})
bmin

C 1-3   F 4-8   M 4-8  F 9-13  M 9-13  \
Nutrition                                                                
Energy                          1000.0  1200.0  1400.0  1600.0  1800.0   
Protein                           13.0    19.0    19.0    34.0    34.0   
Fiber, total dietary              14.0    16.8    19.6    22.4    25.2   
Folate, DFE                      150.0   200.0   200.0   300.0   300.0   
Calcium, Ca                      700.0  1000.0  1000.0  1300.0  1300.0   
Carbohydrate, by difference       50.0    50.0    50.0    50.0    50.0   
Iron, Fe                           7.0    10.0    10.0     8.0     8.0   
Magnesium, Mg                     80.0   130.0   130.0   240.0   240.0   
Niacin                             6.0     8.0     8.0    12.0    12.0   
Phosphorus, P                    460.0   500.0   500.0  1250.0  1250.0   
Potassium, K                    3000.0  3800.0  3800.0  4500.0  4500.0   
Riboflavin                         0.5     0.6     0.6     0.9     0.9   
Thiamin                            0.5     0.6     0.6     0.9     0.9   
Vitamin A, RAE                   300.0   400.0   400.0   600.0   600.0   
Vitamin B-12                       0.9     1.2     1.2     1.8     1.8   
Vitamin B-6                        0.5     0.6     0.6     1.0     1.0   
Vitamin C, total ascorbic acid    15.0    25.0    25.0    45.0    45.0   
Vitamin E (alpha-tocopherol)       6.0     7.0     7.0    11.0    11.0   
Vitamin K (phylloquinone)         30.0    55.0    55.0    60.0    60.0   
Zinc, Zn                           3.0     5.0     5.0     8.0     8.0   
Total fat (NLEA)                  22.0    27.0    31.0    36.0    40.0   
Fatty acids, total saturated       3.0     4.0     5.0     5.0     6.0   

                                F 14-18  M 14-18  F 19-30  M 19-30  F 31-50  \
Nutrition                                                                     
Energy                           1800.0   2200.0   2000.0   2400.0   1800.0   
Protein                            46.0     52.0     46.0     56.0     46.0   
Fiber, total dietary               25.2     30.8     28.0     33.6     25.2   
Folate, DFE                       400.0    400.0    400.0    400.0    400.0   
Calcium, Ca                      1300.0   1300.0   1000.0   1000.0   1000.0   
Carbohydrate, by difference        50.0     50.0     50.0     50.0     50.0   
Iron, Fe                           15.0     11.0     18.0      8.0     18.0   
Magnesium, Mg                     360.0    410.0    310.0    400.0    320.0   
Niacin                             14.0     16.0     14.0     16.0     14.0   
Phosphorus, P                    1250.0   1250.0    700.0    700.0    700.0   
Potassium, K                     4700.0   4700.0   4700.0   4700.0   4700.0   
Riboflavin                          1.0      1.3      1.1      1.3      1.1   
Thiamin                             1.0      1.2      1.1      1.2      1.1   
Vitamin A, RAE                    700.0    900.0    700.0    900.0    700.0   
Vitamin B-12                        2.4      2.4      2.4      2.4      2.4   
Vitamin B-6                         1.2      1.3      1.3      1.3      1.3   
Vitamin C, total ascorbic acid     65.0     75.0     75.0     90.0     75.0   
Vitamin E (alpha-tocopherol)       15.0     15.0     15.0     15.0     15.0   
Vitamin K (phylloquinone)          75.0     75.0     90.0    120.0     90.0   
Zinc, Zn                            9.0     11.0      8.0     11.0      8.0   
Total fat (NLEA)                   40.0     49.0     44.0     53.0     40.0   
Fatty acids, total saturated        6.0      7.0      7.0      8.0      6.0   

                                M 31-50   F 51+   M 51+  
Nutrition                                                
Energy                           2200.0  1600.0  2000.0  
Protein                            56.0    46.0    56.0  
Fiber, total dietary               30.8    22.4    28.0  
Folate, DFE                       400.0   400.0   400.0  
Calcium, Ca

## 2. Limiting Sodium, Carbohydrate, and Fat Intake

Sodium intake should not exceed more than 2,300 grams per day...for anyone! The US Dietary guidelines outlines maximum sodium intake for each demographic by age and sex. Additionally, we add dietary constraints for carbohydrate and fat intake to model diabetic restrictions.

In [1]:
# Load maximum dietary requirements
bmax = pd.read_csv('./diet_maximums.csv').set_index('Nutrition').iloc[:,2:]

# Add dietary contraint for carbohydrates
#bmax.loc[len(bmax.index)] = [70] * len(bmax.columns)
#bmax = bmax.rename({1: 'Carbohydrate, by difference'})

## Uncomment to calculate on your own!

# Add dietary constraint for total fat
#bmax.loc[len(bmax.index)] = np.around(bmin.loc['Energy'].values * 0.35 / 9)
#bmax = bmax.rename({len(bmax.index)-1: 'Total fat (NLEA)'})

# Add dietary constraint for saturated fat
#bmax.loc[len(bmax.index)] = np.around(bmin.loc['Energy'].values * 0.07 / 9)
#bmax = bmax.rename({3: 'Fatty acids, total saturated'})
#bmax

NameError: name 'pd' is not defined

In [36]:
# Concatenate updated minimum and maximum dietary requirements adjusted for individuals with diabetes

b = pd.concat([bmin, -bmax])
b

C 1-3   F 4-8   M 4-8  F 9-13  M 9-13  \
Nutrition                                                                
Energy                          1000.0  1200.0  1400.0  1600.0  1800.0   
Protein                           13.0    19.0    19.0    34.0    34.0   
Fiber, total dietary              14.0    16.8    19.6    22.4    25.2   
Folate, DFE                      150.0   200.0   200.0   300.0   300.0   
Calcium, Ca                      700.0  1000.0  1000.0  1300.0  1300.0   
Carbohydrate, by difference       50.0    50.0    50.0    50.0    50.0   
Iron, Fe                           7.0    10.0    10.0     8.0     8.0   
Magnesium, Mg                     80.0   130.0   130.0   240.0   240.0   
Niacin                             6.0     8.0     8.0    12.0    12.0   
Phosphorus, P                    460.0   500.0   500.0  1250.0  1250.0   
Potassium, K                    3000.0  3800.0  3800.0  4500.0  4500.0   
Riboflavin                         0.5     0.6     0.6     0.9     0.9   
Thiamin                            0.5     0.6     0.6     0.9     0.9   
Vitamin A, RAE                   300.0   400.0   400.0   600.0   600.0   
Vitamin B-12                       0.9     1.2     1.2     1.8     1.8   
Vitamin B-6                        0.5     0.6     0.6     1.0     1.0   
Vitamin C, total ascorbic acid    15.0    25.0    25.0    45.0    45.0   
Vitamin E (alpha-tocopherol)       6.0     7.0     7.0    11.0    11.0   
Vitamin K (phylloquinone)         30.0    55.0    55.0    60.0    60.0   
Zinc, Zn                           3.0     5.0     5.0     8.0     8.0   
Total fat (NLEA)                  22.0    27.0    31.0    36.0    40.0   
Fatty acids, total saturated       3.0     4.0     5.0     5.0     6.0   
Sodium, Na                     -1500.0 -1900.0 -1900.0 -2200.0 -2200.0   

                                F 14-18  M 14-18  F 19-30  M 19-30  F 31-50  \
Nutrition                                                                     
Energy                           1800.0   2200.0   2000.0   2400.0   1800.0   
Protein                            46.0     52.0     46.0     56.0     46.0   
Fiber, total dietary               25.2     30.8     28.0     33.6     25.2   
Folate, DFE                       400.0    400.0    400.0    400.0    400.0   
Calcium, Ca                      1300.0   1300.0   1000.0   1000.0   1000.0   
Carbohydrate, by difference        50.0     50.0     50.0     50.0     50.0   
Iron, Fe                           15.0     11.0     18.0      8.0     18.0   
Magnesium, Mg                     360.0    410.0    310.0    400.0    320.0   
Niacin                             14.0     16.0     14.0     16.0     14.0   
Phosphorus, P                    1250.0   1250.0    700.0    700.0    700.0   
Potassium, K                     4700.0   4700.0   4700.0   4700.0   4700.0   
Riboflavin                          1.0      1.3      1.1      1.3      1.1   
Thiamin                             1.0      1.2      1.1      1.2      1.1   
Vitamin A, RAE                    700.0    900.0    700.0    900.0    700.0   
Vitamin B-12                        2.4      2.4      2.4      2.4      2.4   
Vitamin B-6                         1.2      1.3      1.3      1.3      1.3   
Vitamin C, total ascorbic acid     65.0     75.0     75.0     90.0     75.0   
Vitamin E (alpha-tocopherol)       15.0     15.0     15.0     15.0     15.0   
Vitamin K (phylloquinone)          75.0     75.0     90.0    120.0     90.0   
Zinc, Zn                            9.0     11.0      8.0     11.0      8.0   
Total fat (NLEA)                   40.0     49.0     44.0     53.0     40.0   
Fatty acids, total saturated        6.0      7.0      7.0      8.0      6.0   
Sodium, Na                      -2300.0  -2300.0  -2300.0  -2300.0  -2300.0   

                                M 31-50   F 51+   M 51+  
Nutrition                                                
Energy                           2200.0  1600.0  2000.0  
Protein                         

## Putting it together
*Taken from class material:*

Here we take the different pieces of the puzzle we’ve developed and put them together in the form of a linear program we can solve. Recall that the mathematical problem we’re trying to solve is$$
    \min_x p'x
$$such that$$
     Ax \geq b
$$If we buy a bag of groceries with quantities given by $x$, the total cost of the bag of groceries is the inner product of prices and quantities. Since we’ve converted our units above, this gives us a vector of prices where quantities are all in 100 g or ml units.

The following code block defines a function

In [154]:
from  scipy.optimize import linprog as lp
import numpy as np

def solve_subsistence_problem(FoodNutrients,Prices,diet_min,diet_max,tol=1e-6):
    """Solve Stigler's Subsistence Cost Problem.

    Inputs:
       - FoodNutrients : A pd.DataFrame with rows corresponding to foods, columns to nutrients.
       - Prices : A pd.Series of prices for different foods
       - diet_min : A pd.Series of DRIs, with index corresponding to columns of FoodNutrients,
                    describing minimum intakes.
       - diet_max : A pd.Series of DRIs, with index corresponding to columns of FoodNutrients,
                    describing maximum intakes.
       - tol : Solution values smaller than this in absolute value treated as zeros.
       
    """
    p = Prices.apply(lambda x:x.magnitude).dropna()

    # Compile list that we have both prices and nutritional info for; drop if either missing
    use = list(set(p.index.tolist()).intersection(FoodNutrients.columns.tolist()))
    p = p[use]

    # Drop nutritional information for foods we don't know the price of,
    # and replace missing nutrients with zeros.
    Aall = FoodNutrients[p.index].fillna(0)

    # Drop rows of A that we don't have constraints for.
    Amin = Aall.loc[diet_min.index]

    Amax = Aall.loc[diet_max.index]

    # Minimum requirements involve multiplying constraint by -1 to make <=.
    A = pd.concat([Amin,Amax])

    b = pd.concat([diet_min, -diet_max]) # Note sign change for max constraints

    # Now solve problem!  (Note that the linear program solver we'll use assumes
    # "less-than-or-equal" constraints.  We can switch back and forth by
    # multiplying $A$ and $b$ by $-1$.)

    result = lp(p, A, b, method='interior-point')

    result.A = A
    result.b = b
    result.diet = pd.Series(result.x,index=p.index)

    return result

## Using =solve_subsistence_problem= to analyze diet
Let’s choose a particular group (type of person with particular dietary requirements) and solve the subsistence problem for them. For this project, we studied males and females ages 19-39 as this encompasses most undergraduate students at UC Berkeley.

## Minimum Cost Diet: Males ages 19-30

In [155]:
group = 'M 19-30'
tol = 1e-6

result = solve_subsistence_problem(FoodNutrients,Prices,bmin[group],bmax[group],tol=tol)

print("Cost of diet for %s is $%4.2f per day.\n" % (group,result.fun))

# Put back into nice series
diet = result.diet

print("\nDiet (in 100s of grams or milliliters):")
print(diet[diet >= tol])  # Drop items with quantities less than precision of calculation.
print()

tab = pd.DataFrame({"Outcome":np.abs(result.A).dot(diet),"Recommendation":np.abs(result.b)})
print("\nWith the following nutritional outcomes of interest:")
print(tab)
print()

print("\nConstraining nutrients are:")
excess = tab.diff(axis=1).iloc[:,1]
print(excess.loc[np.abs(excess) < tol*100].index.tolist())

Cost of diet for M 19-30 is $7.93 per day.


Diet (in 100s of grams or milliliters):
Food
Organic rice                                      0.164185
Lean ground turkey                                0.235346
Rotisserie Chicken                                0.107124
Salted shrimp                                     0.204478
Broccoli Florets                                  0.226299
Challenge unsalted butter                         0.469688
Org large mushrooms                               0.781906
Ottogi Gold Mayonnaise                            0.133122
Sajo Light Tuna                                   0.134542
ground beef                                       0.337050
pork belly skinless thin                          0.116475
cheddar cheese                                    0.040182
Dole, Organic Bananas, 2 lb Bag                   0.078578
Sanuki Udon                                       0.298911
Wonderful Halos Mandarins, 3lb Bag                0.776875
Org Steel cut oats       

## Minimum Cost Diet: Females ages 19-30

In [156]:
group = 'F 19-30'
tol = 1e-6

result = solve_subsistence_problem(FoodNutrients,Prices,bmin[group],bmax[group],tol=tol)

print("Cost of diet for %s is $%4.2f per day.\n" % (group,result.fun))

# Put back into nice series
diet = result.diet

print("\nDiet (in 100s of grams or milliliters):")
print(diet[diet >= tol])  # Drop items with quantities less than precision of calculation.
print()

tab = pd.DataFrame({"Outcome":np.abs(result.A).dot(diet),"Recommendation":np.abs(result.b)})
print("\nWith the following nutritional outcomes of interest:")
print(tab)
print()

print("\nConstraining nutrients are:")
excess = tab.diff(axis=1).iloc[:,1]
print(excess.loc[np.abs(excess) < tol*100].index.tolist())

Cost of diet for F 19-30 is $6.12 per day.


Diet (in 100s of grams or milliliters):
Food
Organic rice                                      0.088993
Lean ground turkey                                0.060125
Rotisserie Chicken                                0.146762
Salted shrimp                                     0.111676
Broccoli Florets                                  0.192399
Challenge unsalted butter                         0.077412
Org large mushrooms                               0.487488
Ottogi Gold Mayonnaise                            0.097864
Sajo Light Tuna                                   0.025349
ground beef                                       0.069743
pork belly skinless thin                          0.074253
cheddar cheese                                    0.047497
Dole, Organic Bananas, 2 lb Bag                   0.064859
Sanuki Udon                                       0.163491
Wonderful Halos Mandarins, 3lb Bag                0.973186
Org Steel cut oats       

## Effects of Price Changes on Subsistence Diet Cost

If the price of a particular kind of food changes, how much does the price of the diet change? The code below creates a graph which changes prices away from the `base’ case one food at a time, and plots changes in total diet cost.

In [157]:
import cufflinks as cf
cf.go_offline()

scale = [.5,.6,.7,.8,.9,1.,1.1,1.2,1.3,1.4,1.5]

cost0 = solve_subsistence_problem(FoodNutrients,Prices,bmin[group],bmax[group],tol=tol).fun

Price_response={}
for s in scale:
    cost = {}
    for i,p in enumerate(Prices):
        my_p = Prices.copy()
        my_p[i] = p*s
        result = solve_subsistence_problem(FoodNutrients,my_p, bmin[group], bmax[group],tol=tol)
        cost[Prices.index[i]] = np.log(result.fun/cost0)
    Price_response[np.log(s)] = cost

Price_response = pd.DataFrame(Price_response).T
Price_response.iplot(xTitle='log price change',yTitle='log cost change')

The composition of food sources in our diet are generally non-seasonal, demand elastic, and sufficiently supplied. 

If we examine the impact on prices by seasonal changes and the pandemic, an increase in price of a vegetable can lead to the options replacing with other substitutes. As shown in the graph, a log change of 0.4 in price of all food sources result in a fairly small log change in the cost of our diet. Overall, our minimal cost diet is relatively non-sensitive to price changes, as the subsistence problem solver function can adjust the diet accordingly to satisfy nutritional needs.


## Limitations

When analyzing the *minimum cost diet*, it is important to note that included food items are unlikely to be available for sale in given quantities. The actual price of a minimum cost diet may be more than what is calculated taking this into consideration. Additional research such as how items are packaged and in what quantity must be made to definitively answer the question of subsidence diets. 

## Conclusion

In conclusion, the minimum cost diets for males and females age 19-30 with diabetes at UC Berkeley is $7.93 and $6.12, respectively. This diet takes into account the dietary needs of those with diabetes--which includes the necessity of reduced carbohydrates, sodium, and fats intake--in addition to the foods that are accessible and typically consumed from nearby grocery stores. The differences in prices may be attributed to the differences in recommended calories for selected groups, however, the constraints remain the same for both. Our analysis does not include how some items are sold in fixed quantities, therefore the real minimum cost diet may be priced higher. Ultimately, young adults with diabetes at UC Berkeley have a large selection of food items to maintain a healthy diet at a reasonable price and are encouraged to try different variations of recipes to suit their tastes!